### ISSUE TITLE
Cover hurdle outer grid is a dense 840-pt tensor with ESS=1 (most points are wasted full-data inner solves)

### ISSUE THREAD

**Issue by gcol33** [gcol33/tulpaObs#20]
## Summary

The cover-hurdle joint nested-Laplace outer grid is a dense tensor product whose
posterior collapses onto ~1 point, so the overwhelming majority of grid points
run a full inner Newton solve for negligible posterior weight.

## Measured (MOT Plantago lanceolata cover fit)

1.16M plot rows, 412 cells, `n.threads = 12`, `method = "nested_laplace"`,
lognormal positive arm. Wall time **56.2 min**. From the saved fit's
`$joint$theta_grid` / `$joint$weights`:

- Grid = full tensor `sigma(5) x rho(4) x alpha(6) x phi_pos(7)` = **840 points**.
- **Effective sample size = 1.0 of 840** (top-1 weight 0.999; only **2** points
  with weight >= 1e-6 of total; 1 point with weight >= 1e-3).
- Inner Newton `n_iter` per grid point: 8-50, mean ~11. So ~838 points each run
  a full inner solve over 1.16M rows for ~zero posterior contribution.
- Posterior concentrates at `sigma=0.55, rho=0.2, alpha=1e-4 (~0), phi=0.98` --
  e.g. all 5 nonzero `alpha` layers and most `phi` layers are wasted.

## Where

`R/family_cover_hurdle.R` builds the per-axis grids (`sigma_grid`, `rho_grid`,
`sigma_pos_grid`/alpha, `phi_grid`) and passes them to
`tulpa::tulpa_nested_laplace_joint()`, which expands them to the full tensor.
`adaptive_grid = TRUE` (default) densifies near the mode but does not prune the
dense base tensor.

## Proposal (any of)

1. **CCD design** instead of the full tensor for the outer grid -- the
   `occu_cover()` `joint_coupled` path is ~10x faster on comparable structure;
   `gcol33/tulpa#33` (generic outer-grid helper / CCD) is the target.
2. **Weight pruning**: compute a cheap log-marginal pre-estimate per base-grid
   point (e.g. at a warm-started mode without full convergence) and skip the
   full inner solve where it is negligible relative to the running max.
3. At minimum, **coarsen** the `alpha` and `phi` axes (currently 6x7 = 42 with
   posterior mass on ~1); a `3x3x3x3` base (81 pts) would still bracket the mode
   and cut the fit ~10x.

## Note

56 min is acceptable for the current deliverable (confirmed with the
collaborator), so this is filed as a **performance inefficiency**, not a blocker.
The wasted-solve fraction (~99.8%) is the actionable signal. Found during the
cover-hurdle runtime investigation; measured fit at
`cover-glaser/validation/runs/cover_Plantago_lanceolata_2026-05-30/`.

**Comment by gcol33** [comment]
Fixed via weight-pruning (the principled option from the issue). The cover-hurdle joint path (R/family_cover_hurdle.R) now forwards `prune` / `prune.tol` (default TRUE / 1e-4) into the joint_control passed to tulpa::tulpa_nested_laplace_joint(). The engine-side cheap-pass prune (tulpa src/nested_laplace_grid.h) solves the pilot cell, evaluates a cheap Gaussian-Laplace screening log-marginal at the pilot mode for every other base-grid cell, softmax-normalises, and SKIPS the full inner Newton on cells whose screened weight is < prune_tol of the running max (the pilot is never pruned). It returns prune_n_pruned / prune_mask. The keys are declared in the cover family's control_keys (R/obs_families.R) so validation accepts them.

This is a perf change, NOT a correctness change -- verified on the issue's dense-tensor structure (sigma(5) x rho(4) x sigma_pos(6) x phi(7) = 840 cells, beta arm): 831 of 840 cells pruned (98.9%), wall 5.9s -> 2.4s (2.43x) on a small N=400 example (the speedup scales with inner-solve cost, i.e. far larger on the 1.16M-row fit), and the answer is preserved -- beta_occ max|diff| 6e-4, beta_pos max|diff| 1.7e-3, alpha posterior mean 1.3247 (unpruned) vs 1.3193 (pruned).

Test (tests/testthat/test-cover-hurdle-adaptive-grid.R, 'outer-grid pruning keeps the mode and leaves estimates unchanged'): asserts >50% of cells pruned, modal hyperparameters agree, and pruned-vs-unpruned coefficient estimates match within tolerance. Passing under NOT_CRAN.

**Comment by gcol33** [comment]
Reopening: the cheap-pass prune that closed this (`prune = TRUE, prune_tol = 1e-4`) is fast but **returns the wrong posterior** on the real 1.16M-row BYM2 cover fit. The cheap screen (one Newton step from a global pilot) mis-ranks cells by O(1e5) log-units and prunes the true mode `[sigma=0.55, rho=0.20, alpha=0]` (full-grid weight 0.999); it keeps `[sigma=0.10, rho=0.95, alpha=0.10]` (true weight 2e-321) and reports it with weight 1.0. Wall-clock 56.2 -> 10.9 min, but the BYM2 variance partition inverts (rho 0.20 -> 0.95) so the maps would be wrong. Root cause + reproducer in gcol33/tulpa#43. The grid-waste observation here still stands — but the safe fix is a CCD/coarser grid where every retained cell is fully solved, or a rank-faithful screen (warm-start from nearest solved neighbour). Until then, the cover hurdle should not default `prune = TRUE`.

**Comment by gcol33** [comment]
## Resolved

### Default change

The cover-hurdle joint nested-Laplace path (`R/family_cover_hurdle.R`) **no longer defaults `prune = TRUE`**. The cheap-pass prune was the mechanism this issue was closed with, but as filed in gcol33/tulpa#43 the screen as implemented mis-ranked cells and could drop the true posterior mode, so a fast-but-wrong default is not acceptable. The default is now `prune = FALSE` (the correct full grid).

```r
prune = control$prune %||% FALSE   # was %||% TRUE
```

### Rank-safe speed path (this issue's option 3)

The default speed path is the **adaptive grid** (`adaptive_grid = TRUE`, already on): it brackets the mode with FULL inner solves over a coarser base tensor and densifies near the mode, so it **never approximates the marginal** and cannot drop the true mode. This is the rank-safe variant of option 3. The 56 min full-grid run remains acceptable for the current deliverable (confirmed with the collaborator), so correctness is the right default.

### Cheap-pass prune is now opt-in and safe

`control$prune = TRUE` is still available. Upstream (gcol33/tulpa#43) the cheap pass was rewritten to be rank-faithful (a neighbour-warm-started lattice sweep keeps every cheap mode near its cell's true mode) and gated (a safety check warns and falls back to the full grid if the screen's ranking looks unreliable). So even opt-in, a silently-wrong pruned posterior is impossible.

### Verification

`test-cover-hurdle-adaptive-grid.R` "outer-grid pruning keeps the mode and leaves estimates unchanged" passes: with the rewritten screen the cover hurdle prunes >50% of the grid while the modal hyperparameters and beta estimates match the un-pruned dense-grid fit (tolerances 0.02-0.05), and the safety-gate fallback does NOT trigger (the screen is rank-faithful on this workload).

The grid-waste observation in this issue still holds (ESS ~ 1 of 840); the actionable change is that the *correct* default is the full / adaptive grid, with the (now rank-faithful + gated) cheap-pass prune as an opt-in.
